### ISSUE TITLE
Create timestamp-query2 for tiled architectures

### ISSUE THREAD

**Issue by greggman** [gpuweb/gpuweb#4995]
The `'timestamp-query'` query feature seems unportable (unless it's just a bug)

For example, [this example](https://jsgist.org/?src=99980fb6b01956fda772056ef603508f) does 9 render passes. Each pass is timed with timestamps. Each pass is drawing a bunch of cube instances with no depth test, no culling, and blending on so effectively, every pixel is required to be rendered. The number of instances per pass changes over time from 1 to 3000.

The expectation is that less instances take less time, more instances take more time and on Windows 11 with NVidia that's what I see

```
0:   3015µs numInst:  1646
1:    655µs numInst:   323
2:     66µs numInst:    82
3:   1376µs numInst:  1146
4:   3146µs numInst:  2535
5:   3867µs numInst:  2973
6:   3736µs numInst:  2057
7:   1376µs numInst:   629
8:     66µs numInst:     2
```

But on M1 Mac that's not what I see. Instead, the last pass takes the most time

```
0:   1835µs numInst:  1689
1:   2294µs numInst:   350
2:   2294µs numInst:    69
3:   3211µs numInst:  1104
4:   5439µs numInst:  2503
5:   8258µs numInst:  2981
6:  10420µs numInst:  2097
7:  11207µs numInst:   665
8:  11010µs numInst:     1
```

I'm guessing the issue is that on M1 (and other tiled hardware?), the passes are being combined into one pass so the timing of each pass has the same start time but different end times ... or something along those lines.

This makes `'timestamp-query'` on these devices problematic. You need to somehow guess what kind of hardware you're on and then adjust your code appropriately. Should `'timestamp-query'`  work the same everywhere it's available? Assuming that it can not work the same everywhere, should it not be available on those devices where it behaves like this? And, if possible a different `'timestamp-query-xxx'` feature be specified for these devices? That way, at least the developer can know how to use the results?

A confounding issue is compute. At least in my tests, `'timestamp-query'` with compute passes behaves as expected on these devices (a pass doing more work takes more time), [example](https://jsgist.org/?src=21f79a052fde4c1b49de76b3e81594df). So maybe we'd need to split features into `'timestamp-query-compute'` and `'timestamp-query-renderpass-combined'` for these devices that don't support `'timestamp-query'` in render passes as separate timings?

In other words, devices where each compute and render pass gets its own timestamps advertize `'timestamp-query'`. devices were render passes are combined and compute passes are not advertise `'timestamp-query-renderpass-combined'` and `'timestamp-query-compute'`

or maybe there are other solutions.

I think some might suggest two other solutions

* ignore this issue - timing isn't important
* Update the specs to say how it works is even more undefined. 
 
  note: the spec already says the values are implementation-defined. maybe that's enough. It seems strange to have such a feature though. If it's going to stay that way, maybe it should be moved to a developer only feature as it's not portable.

**Comment by Kangz** [comment]
While I understand the concern, and we would like timestamp queries to be as useful as possible, tilers, parallelism on non-tilers and a lot of other factors mean that using timestamps come with a lot of caveats. We should be clearer about them in the spec, but I don't think we can ever paper over all the caveats.

Also in the example you described it seems that the same render pass (one the same attachments) is started multiple times, which ideally applications shouldn't do for performance. Metal does an optimization that's making timestamps weird here but also applications shouldn't do the same render pass multiple times in a row, unless they need to break the usage scopes for some reason.

IDK that increasing the feature soup will make things a lot clearer for developers, they would have to write more code to get timestamps to work everywhere but will they know what `combined` means? (also it seems to be an implementation detail of Metal only on Apple GPUs?)

**Comment by mwyrzykowski** [comment]
<img width="364" alt="Screenshot 2024-12-05 at 3 51 41 PM" src="https://github.com/user-attachments/assets/029faded-8d5c-4970-b3c4-f3f992525ea1">


fwiw @greggman on an M2 Mac I see the top results on Safari Tech Preview 208, but on Chrome Version 133.0.6878.0 I see the bottom which is closer to what is reported on M1

**Comment by mwyrzykowski** [comment]
I would second what @Kangz said and take no action here, the sample is creating N passes within a single command buffer from my reading. There is no guarantee on TBDR at least Apple's GPUs that the pass 1 completes before pass N, pass N may complete first.

**Comment by greggman** [comment]
Can we document this in some useful way or do we just leave it as "timestamp-query" is UB period?

**Comment by Kangz** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/GPU-Web-2024-12-04#create-timestamp-query2-for-tiled-architectures-4995">GPU Web WG 2024-12-04 Atlantic-time</a></summary>

* KN: attempting overview but this is gregg's issue
* KG: spec says it's implementation-defined. This is expected, 
* JB: Agree
* AB: Two uses for timestamps that need to be considered. One is for adaptive performance, for dynamically scaling down resolution or LoD. Currently don't think we have anything for that, since we only have timestamps on per pass, not begin/end of frame. Second is for profiling. For that, it gets really complicated. Can go into some details about why it's complicated on ARM.
* KG: Think we have enough of an idea of ways it's complicated, so not necessary.
* KG: For total frame timings, thought you can enqueue an empty pass that just gives you timing.
* AB: Ah right.
* AB: To make timestamps make sense on ARM, only way to do it is with full pipeline barriers between every pass
* KN: Unfortunately no way to insert full pipeline barriers. But can use some workaround like wait for GPU idle, then submit only that work, and time it. Slow but gives results
* 

</details>
